platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,942 @@
+- It's a very [complicated interface](#steep-learning-curve)
+
+**Building a form from a schema**
+- The schema to build the form fields must sometimes differ from the validation
+  schema, which means data transformation functions are required
+  - The 21-526EZ form is a prime example of this because of its complexity
+- Again, the [learning curve](#steep-learning-curve) gets in the way of rapid
+  progress on more complicated forms
+  - Engineers have to learn the interface for making dynamic changes through
+    hooks provided in the `uiSchema` instead of using standard React conventions
+- Using external services to determine which fields to render is cumbersome
+  - Rather than conditionally rendering the fields, engineers must generate a
+    schema
+- Paring `schema` and `uiSchema` makes sharing reusable patterns awkward
+  - Implementation details such as the data paths are baked into the
+    definitions of these reusable patterns
+
+### High Level Design
+Formulate will be a modular library consisting of helper functions and
+components to aid in the rapid development of forms on VA.gov. These functions
+and components can be conceptualized as **sub-modules** for:
+- [Building form pages](#form-page-builder)
+  - Wrappers for Formik components using the design system
+  - Re-exporting necessary Formik components
+- [Routing](#routing)
+  - Simple set-up for using Formik across multiple pages
+  - Components for rendering the standard form "chrome"
+  - Looping through a series of pages per item in an array
+  - Conditionally-rendered pages
+- [Save in progress](#save-in-progress)
+  - Manual save
+  - Auto save
+  - Data migrations
+
+**Each sub-module will be independent.** If a form doesn't need the multi-page
+routing, the team building it simply won't use the routing sub-module, and it
+won't be included in their application bundle. Likewise, if a form needs more
+complex routing, the team building it can opt to roll their own, but still reap
+the benefits of using the rest of the system.
+
+_Note: "sub-modules" here refer to independent parts of a larger system and have
+nothing to do with [Git
+Sub-Modules](https://git-scm.com/book/en/v2/Git-Tools-Sub-Modules)._
+
+Additionally, the library will provide a number of helper functions and
+components for:
+- Field-level schema validation
+- [Building a review page](#json-schema-validation)
+
+## Specifics
+
+### Detailed Design
+**Key technologies** we'll use:
+- React
+- TypeScript
+- [`api-extractor`](https://api-extractor.com/pages/overview/intro/)
+- [React router](https://reactrouter.com/)
+- [Formik](https://formik.org/)
+
+The following sections outline the various parts of the system.
+
+#### Form page builder
+Formulate will leverage Formik for much of the form page building. It will
+**re-export Formik components** where it makes sense and **provide light
+wrappers for Formik input fields** using design system components.
```

Comment:
I would like to see more about how validation is handled for form page components. The current form system implements a particular pattern where validation is run instantly and shown depending on the visited/blurred state of the field. And when you submit the form, all those errors are revealed and the first one is focused on. Formik provides the hooks needed to do all that (I think), but in this doc I'm not sure where that would live (some of it would live in the form field wrappers, but not all of it can go in there). I'd also be interested to know how we should write validation rules, for both regular form fields and also dynamic fields. Should we use Yup? Something else? I know jsonschema is mentioned below, but it's in the context of the final submission.

Making validation "just work" is I think one of the big strengths of the current form system, since it's very easy to mess up and create inaccessible flows. So I think Formulate should make that priority as well.

Reply:
Ah, great call out. I'll go back over and strengthen that section for handling validation, outlining the specifics for when validation will run and how focus will be handled. :+1: